File: AoC_2022/day4/main.go
package main

import (
	utils "AoC_Golang/myutils"
	"fmt"
	"strconv"
	"strings"
)

func main() {
	fmt.Println("Welcome to day 4")
	data := utils.ReadStrings("input.txt")
	part1(data)
	part2(data)

}

func part1(data []string) {
	total := 0

	for _, line := range data {
		nums := strings.Split(line, ",")
		left := strings.Split(nums[0], "-")
		right := strings.Split(nums[1], "-")
		l1, _ := strconv.Atoi(left[0])
		l2, _ := strconv.Atoi(left[1])
		r1, _ := strconv.Atoi(right[0])
		r2, _ := strconv.Atoi(right[1])

		if ((l1 <= r1) && (l2 >= r2)) || ((r1 <= l1) && (r2 >= l2)) {
			total++
		}
	}
	fmt.Printf("Part 1 : %d\n", total)
}

func part2(data []string) {
	total := 0

	for _, line := range data {
		nums := strings.Split(line, ",")
		left := strings.Split(nums[0], "-")
		right := strings.Split(nums[1], "-")
		l1, _ := strconv.Atoi(left[0])
		l2, _ := strconv.Atoi(left[1])
		r1, _ := strconv.Atoi(right[0])
		r2, _ := strconv.Atoi(right[1])

		if (l1 == r1) || (l2 == r2) || (l1 == r2) || (l2 == r1) {
			total++
		} else if (l1 <= r1 && l2 >= r1) || (r1 <= l1 && r2 >= l1) {
			total++
		}

	}
	fmt.Printf("Part 2 : %d\n", total)
}

File: AoC_2022/day4/main.py
# lines = open("input.txt").read().splitlines()

# s = 0
# for line in lines:
#     l1, l2 = line.split(",")
#     (min1, max1), (min2, max2) = l1.split("-"), l2.split("-")
#     min1, max1, min2, max2 = list(map(int, [min1, max1, min2, max2]))
#     # print(l1, l2)
#     # print(min1, max1, min2, max2)
#     if (min1 <= min2 <= max1) or (min2 <= min1 <= max2):
#         s += 1


# print(s)

import re

data = open("input.txt").read()

lines = data.splitlines()

part1 = 0
part2 = 0
for line in lines:
    a, b, c, d = list(int(i) for i in re.findall(r'\d+', line))
    e, f = set(range(a, b+1)), set(range(c, d+1))
    if e <= f or f <= e:
        part1 += 1
    if e & f:
        part2 += 1

print(part1, part2)


print(sum(1 for string in open("input.txt", "r").read().split('\n') if ((int(string.split(',')[0].split('-')[0]) >= int(string.split(',')[1].split('-')[0]) and int(string.split(',')[0].split('-')[1]) <= int(string.split(',')[1].split('-')[1])) or (int(string.split(',')[1].split('-')[0]) >= int(string.split(',')[0].split('-')[0]) and int(string.split(',')[1].split('-')[1]) <= int(string.split(',')[0].split('-')[1])))))